### DISCUSSION TITLE
version 1, 2, 3... supported versions, end of life? Can you explain?

### DISCUSSION THREAD

**Discussion by coclav** [PHPOffice/PhpSpreadsheet#4196] [General]
Good morning, 

I see there are currently 3 major versions being released / maintained (3.3.0, 2.3.0 and 1.29.2 as i'm writing)

We are currently on version 1. I'm looking for the following information

- How long will v1 be maintained ? Is there an end of life to it ? 
- Is there an upgrade guide to v2 and / or v3 ?
- what are the main feature difference from 1 / 2 / 3

Please help us understand the versioning approach?

I have looked in the documentation unfortunately I can't find any answers.

Thank you

**Comment by oleibman** [comment]
We do all new development on the master branch, which is now version 3. Some security exposures which came to our attention just before release 3 were, of course, fixed for that release. We back-ported those changes to versions 1 and 2, leading to the situation you now see. We do not have an end-of-life date for those versions. Our plan, at least for now, is to continue to back-port security fixes as they arise. We are unlikely to back-port other changes.

The changes in each release, both major and minor, are in the master version of CHANGELOG.md. For Release 2 (search for 2.0.0), the major change was more-or-less internal, adding native typing in place of doc-block typing whenever possible. For Release 3 (search for 3.3.0), the major change was dynamic arrays.

It is expected that most applications will not need to change if you migrate to a newer release. You will, of course, have to account for deprecations that have been removed, and for behavior changes documented under "Changes" for each release.

**Reply by coclav** [reply]
Thanks for the clear explanation.
